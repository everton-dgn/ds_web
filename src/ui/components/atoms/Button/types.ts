import { ButtonHTMLAttributes, ReactNode } from 'react'

export type SizeProps = 'small' | 'medium' | 'large' | 'xLarge'

export type ColorProps = 'gray' | 'blue' | 'red'

export type BtnProps = {
  icon?: ReactNode
  text?: string
  size: SizeProps
  color: ColorProps
  fullWidth?: boolean
} & ButtonHTMLAttributes<HTMLButtonElement>
