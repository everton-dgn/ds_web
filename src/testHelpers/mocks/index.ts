export * from './matchMedia'
export * from './reactRouterDom'
