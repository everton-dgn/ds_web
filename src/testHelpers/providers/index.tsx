export * from './providerComponents'
export * from './providerHooks'
