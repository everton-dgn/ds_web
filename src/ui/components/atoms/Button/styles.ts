import styled, { css } from 'styled-components'
import T from 'ui/theme'
import { BtnProps } from './types'

const SizesBtn = {
  small: css`
    height: min-content;
    font-size: ${T.fonts.sizes.xSmall};
    padding: ${T.spacings.xSmall} ${T.spacings.small};
  `,
  medium: css`
    height: min-content;
    font-size: ${T.fonts.sizes.small};
    padding: ${T.spacings.small} ${T.spacings.xLarge};
  `,
  large: css`
    height: min-content;
    font-size: ${T.fonts.sizes.small};
    padding: ${T.spacings.medium} ${T.spacings.xxxLarge};
  `,
  xLarge: css`
    height: min-content;
    font-size: ${T.fonts.sizes.small};
    padding: ${T.spacings.medium} ${T.spacings.xtreme};
  `
}

const ColorsBtn = {
  gray: css`
    background-color: ${T.colors.gray};

    &:hover {
      background-color: ${T.colors.grayHover};
    }
  `,
  blue: css`
    background-color: ${T.colors.primary};

    &:hover {
      background-color: ${T.colors.primary};
    }
  `,
  red: css`
    background-color: ${T.colors.red};

    &:hover {
      background-color: ${T.colors.redHover};
    }
  `
}

export const Container = styled.button<BtnProps>`
  ${({ fullWidth, color, size }) => css`
    cursor: pointer;
    border-radius: ${T.border.radius};
    font-weight: 400;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    flex-wrap: nowrap;
    border: 2px solid transparent;
    text-shadow: 0 0 3px #00000099;
    transition: background-color 0.15s ease-in-out;
    width: 100%;
    ${!!color && ColorsBtn[color]};
    ${!!size && SizesBtn[size]}

    ${T.breakpoints.custom(480)} {
      width: ${fullWidth ? '100%' : 'fit-content'};
      min-width: ${fullWidth ? '300px' : 'inherit'};
    }

    &:disabled,
    &:disabled:hover {
      opacity: 0.5;
      cursor: not-allowed;
      background-color: ${T.colors.disabled};
    }

    &:focus {
      border: 2px solid #fff;
      box-shadow: 0 0 0 2px #222;
    }

    span {
      color: ${T.colors.white};
    }

    svg + span {
      margin-left: 10px;
    }

    svg {
      width: 2rem;
      min-width: 2rem;
      height: 2rem;
      min-height: 2rem;
      color: ${T.colors.white};
    }
  `}
`
