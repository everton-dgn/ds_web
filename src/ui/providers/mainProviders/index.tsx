import { GlobalStyles } from 'ui/globalStyles'
import * as C from 'ui/components'
import { MainProvidersProps } from './types'

export const MainProviders = ({ children }: MainProvidersProps) => (
  <C.ErrorBoundary>
    {children}
    <GlobalStyles />
  </C.ErrorBoundary>
)
