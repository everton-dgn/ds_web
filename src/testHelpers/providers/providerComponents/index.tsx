import { render, RenderResult } from '@testing-library/react'
import { ReactNode } from 'react'
import { MainProviders } from 'ui/providers'

export const renderWithProviders = (children: ReactNode): RenderResult =>
  render(<MainProviders>{children}</MainProviders>)
