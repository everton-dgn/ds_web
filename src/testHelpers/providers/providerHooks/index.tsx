import { renderHook } from '@testing-library/react'
import { MainProviders } from 'ui/providers'
import { RenderHooksProviderProps, WrapperProps } from './types'

const wrapper = ({ children }: WrapperProps) => (
  <MainProviders>{children}</MainProviders>
)

export const renderHooksProvider = (children: RenderHooksProviderProps) =>
  renderHook(children, { wrapper })
