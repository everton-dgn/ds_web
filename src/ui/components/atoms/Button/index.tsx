import { forwardRef, memo } from 'react'
import * as S from './styles'
import { BtnProps } from './types'

const Button = forwardRef<HTMLButtonElement, BtnProps>(
  ({ icon, text, ...props }, ref) => (
    <S.Container {...props} ref={ref}>
      {icon}
      {!!text && <span>{text}</span>}
    </S.Container>
  )
)

export default memo(Button)
