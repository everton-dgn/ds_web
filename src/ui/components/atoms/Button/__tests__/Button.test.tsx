import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { renderWithProviders } from 'testHelpers/providers'
import Button from '..'

describe('[Component] Button', () => {
  it('should call a function once on click on the button', async () => {
    const user = userEvent.setup({ delay: null })
    const onClick = jest.fn()

    renderWithProviders(
      <Button
        onClick={onClick}
        text="Mais Informações"
        color="blue"
        size="large"
      />
    )

    const btn = screen.getByRole('button', { name: /Mais Informações/i })
    await user.click(btn)

    expect(onClick).toHaveBeenCalled()
    expect(onClick).toHaveBeenCalledTimes(1)
  })

  it('should render a small blue button with text', () => {
    const { container } = renderWithProviders(
      <Button text="Mais Informações" color="blue" size="small" />
    )

    const btn = screen.getByRole('button', { name: /Mais Informações/i })
    expect(btn).toHaveStyle({
      height: 'min-content',
      padding: '8px 12px',
      fontSize: '1.2rem'
    })

    expect(btn).toBeEnabled()
    expect(container).toMatchSnapshot()
  })

  it('should render a medium blue button with text', () => {
    renderWithProviders(
      <Button size="medium" text="Mais Informações" color="blue" />
    )

    const btn = screen.getByRole('button', { name: /Mais Informações/i })

    expect(btn).toHaveStyle({
      height: 'min-content',
      padding: '12px 24px',
      fontSize: '1.4rem'
    })
  })

  it('should render a disabled Button', () => {
    renderWithProviders(
      <Button disabled text="Mais Informações" color="blue" size="large" />
    )

    const btn = screen.getByRole('button', { name: /Mais Informações/i })

    expect(btn).toHaveStyle({
      cursor: 'not-allowed',
      opacity: 0.5
    })
    expect(btn).toBeDisabled()
  })

  it('should render a Button with full width', () => {
    renderWithProviders(
      <Button
        disabled
        text="Mais Informações"
        color="blue"
        size="large"
        fullWidth={true}
      />
    )

    const btn = screen.getByRole('button', { name: /Mais Informações/i })

    expect(btn).toHaveStyle({ width: '100%' })
  })
})
