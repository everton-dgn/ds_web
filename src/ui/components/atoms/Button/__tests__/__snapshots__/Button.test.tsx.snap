// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component] Button should render a small blue button with text 1`] = `
.c0 {
  cursor: pointer;
  border-radius: 4px;
  font-weight: 400;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  border: 2px solid transparent;
  text-shadow: 0 0 3px #00000099;
  -webkit-transition: background-color 0.15s ease-in-out;
  transition: background-color 0.15s ease-in-out;
  width: 100%;
  background-color: var(--colors-primary);
  height: -webkit-min-content;
  height: -moz-min-content;
  height: min-content;
  font-size: 1.2rem;
  padding: 8px 12px;
}

.c0:hover {
  background-color: var(--colors-primary);
}

.c0:disabled,
.c0:disabled:hover {
  opacity: 0.5;
  cursor: not-allowed;
  background-color: var(--colors-disabled);
}

.c0:focus {
  border: 2px solid #fff;
  box-shadow: 0 0 0 2px #222;
}

.c0 span {
  color: var(--colors-white);
}

.c0 svg + span {
  margin-left: 10px;
}

.c0 svg {
  width: 2rem;
  min-width: 2rem;
  height: 2rem;
  min-height: 2rem;
  color: var(--colors-white);
}

@media screen and (min-width:480px) {
  .c0 {
    width: -webkit-fit-content;
    width: -moz-fit-content;
    width: fit-content;
    min-width: inherit;
  }
}

<div>
  <button
    class="c0"
    color="blue"
  >
    <span>
      Mais Informações
    </span>
  </button>
</div>
`;
